solidity/contracts/converter/BancorConverter.sol

```diff
@@ -733,11 +801,17 @@ contract BancorConverter is IBancorConverter, SmartTokenController, ContractRegi
         // _amount and the total supply in each reserve from the caller to the converter
         for (uint16 i = 0; i < reserveTokens.length; i++) {
             IERC20Token reserveToken = reserveTokens[i];
-            uint256 reserveBalance = reserveToken.balanceOf(this);
+            bool isETH = reserveToken == IERC20Token(0);
+            uint256 reserveBalance = getReserveBalance(reserveToken) - (isETH ? msg.value : 0);
             uint256 reserveAmount = formula.calculateFundCost(supply, reserveBalance, totalReserveRatio, _amount);
 
             // transfer funds from the caller in the reserve token
-            ensureTransferFrom(reserveToken, msg.sender, this, reserveAmount);
+            if (isETH && msg.value > reserveAmount)
+                msg.sender.transfer(msg.value - reserveAmount);
+            else if (isETH && msg.value < reserveAmount)
+                ensureTransferFrom(IERC20Token(0xc0829421C1d260BD3cB3E0F06cfE2D52db2cE315), msg.sender, this, reserveAmount - msg.value);

```

Comment:
if the converter has ETH reserve, it doesn't know how to handle EtherToken so it won't help much to transfer the EtherToken to the converter (the converter will just have EtherToken balance - not supposed to)
For ETH converters, there are 2 options -
1. Enough ETH was sent with the call - process it and return the remainder (done)
2. NO ETH was sent with the call (require that value == 0) - in this case do the ensureTransferFrom just like any token and then specifically for ETH converter, also call EtherToken.withdraw() to get the ETH from the EtherToken

So if some ETH was sent with the call but not enough, revert.

Reply:
@yudilevi : By `EtherToken.withdraw()`, do you mean `EtherToken.withdrawTo(this, reserveAmount)` (or simply `EtherToken.withdraw(reserveAmount)`)?

Reply:
@yudilevi : By `etherToken.withdraw()`, do you mean `etherToken.withdrawTo(this, reserveAmount)` (or simply `etherToken.withdraw(reserveAmount)`)?

Reply:
@yudilevi : Done.